pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,352 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Lie gradient optimizers"""
+import itertools as it
+import copy
+
+import numpy as np
+import pennylane as qml
+from pennylane.tape import JacobianTape
+from pennylane.transforms import batch_transform
+
+
+
+@qml.qfunc_transform
+def append_time_evolution(tape, hamiltonian, t):
+    r"""Append an approximate time evolution to an existising circuit
+    Trotterize the Hamiltonian and with a single step
+
+    .. math:
+
+        U = \prod_i \exp{-it O_i}
+
+    Then this unitary is appended to the current circuit.
+
+    Args:
+        tape: qml.QuantumTape of circuit
+        hamiltonian: qml.Hamiltonian object
+        t: time evolution size
+
+    Returns:
+
+    """
+    for obj in tape.operations:
+        qml.apply(obj)
+    qml.templates.ApproxTimeEvolution(hamiltonian, t, 1)
+
+    for obj in tape.measurements:
+        qml.apply(obj)
+
+
+class LieGradientOptimizer:
+    r"""Exact Lie gradient optimizer"""
+
+    # pylint: disable=too-few-public-methods
+    pl_paulis = [
+        qml.Identity,
+        qml.PauliX,
+        qml.PauliY,
+        qml.PauliZ,
+    ]
+    pauli_int_to_str = {0: "I", 1: "X", 2: "Y", 3: "Z"}
+    pauli_str_to_int = {"I": 0, "X": 1, "Y": 2, "Z": 3}
+
+    def __init__(self, circuit, hamiltonian, stepsize=0.01, **kwargs):
+        r"""
+        Base class for other gradient-descent-based optimizers.
+
+        A step of the Lie gradient iterates the Lie gradient flow on :math:`\text{SU}(2^N)`.
+        The function to be minimized is :math:`f(U) = \text{Tr}(U \rho_0 U^\dag H)` given a
+        Hamiltonian :math:`H` and initial state :math:`\rho_0`
+
+        .. math::
+
+            U^{(t+1)} = \exp{\epsilon \text{grad}f(U^{(t)}}) U^{(t)}
+
+        where :math:`\epsilon` is a user-defined hyperparameter corresponding to step size.
+
+        The Lie gradient is given by.
+
+        .. math::
+
+             \text{grad}f(U^{(t)}}) = -[U \rho U^\dag, H]
+
+        Subsequent steps of this optimizer will append a Trotterized version of the exact Lie
+        gradient and grow the circuit.
+
+        Args:
+            circuit (qml.QNode): the user-defined hyperparameter :math:`\eta`
+            hamiltonian (qml.Hamiltonian): the user-defined hyperparameter :math:`\eta`
+            stepsize (float): the user-defined hyperparameter :math:`\eta`
+            **kwargs
+        **Examples:**
+
+        >>> @qml.qnode(qml.device("default.qubit", wires=2))
+        ... def quant_fun():
+        ...     qml.RX(0.1, wires=[0])
+        ...     qml.RY(0.5, wires=[1])
+        ...     qml.CNOT(wires=[0,1])
+        ...     qml.RY(0.6, wires=[0])
+        ...     return qml.state()
+        Define a Hamiltonian cost function to minimize.
+        >>> hamiltonian = qml.Hamiltonian(coeffs=[-1.]*3, observables=[qml.PauliX(0), qml.PauliZ(1), qml.PauliY(0)@qml.PauliX(1)])
```

Comment:
Hey @therooler, just asking so that I understand!

So here, we want to minimize `<0|U(x)†HU(x)|0>` where `x` are the ansatz weights, H is some Hamiltonian, and `U` is the unitary we are applying?

- Although the quantum circuit you have above has no trainable parameters, as far as I can tell?

- Could we have the optimizer simply accept a single QNode that ends with `expval(H)`, and extract the Hamiltonian in that form?

E.g., would it make sense to have the optimizer accept
```python
@qml.qnode(dev)
def circuit():
    qml.RX(0.1, wires=[0])
    qml.RY(0.5, wires=[1])
    qml.CNOT(wires=[0,1])
    qml.RY(0.6, wires=[0])
    return qml.expval(H)
```

and the optimizer then:

1. extracts H
2. creates a new QNode that ends in `qml.state()` for internal use?

I'm just wondering if this is a 'less surprising' way for a user to construct such a cost function.

Reply:
There are no parameters that are being trained. The initial circuit is simply the initial state from which we start the optimization.

I didnt know that I could use qml.expval and Hamiltonian together like this, but yes, this would be a more natural approach for sure! Let me try to make this work. This would probably fix the issues I'm having with qml.map as well.

Reply:
> There are no parameters that are being trained. The initial circuit is simply the initial state from which we start the optimization.

Got it!